perf-tool/src/server.cpp

```diff
@@ -109,133 +141,188 @@ void Server::handleServer()
+            int pollTimeMillis = ServerConstants::POLL_INTERVALS;
+            if (!delayedCommands.empty())
+            {
+                int nextOperationDelay = executeAllDueCommands();
+                if (nextOperationDelay > 0)
+                    pollTimeMillis = std::min(nextOperationDelay*1000, pollTimeMillis);
+            }
+            int pollResult = poll(pollFds, activeNetworkClients + ServerConstants::BASE_POLLS, pollTimeMillis);
+            if (pollResult == -1)
             {
                 perror("ERROR on polling, server will now shut down");
                 break;
             }
 
-            if (activeNetworkClients < ServerConstants::NUM_CLIENTS && (pollFds[0].revents & POLLIN))
+            if (pollResult > 0) /* data is ready on at least one socket */
             {
-                /* The accept() call actually accepts an incoming connection */
-                clilen = sizeof(cli_addr);
-
-                /* This accept() function will write the connecting client's address info 
-                 *into the the address structure and the size of that structure is clilen.
-                 */
-                newsocketFd = accept(serverSocketFd, (struct sockaddr *)&cli_addr, &clilen);
-                if (newsocketFd < 0)
-                    perror("ERROR on accept");
-                else
-                    networkClients[activeNetworkClients] = new NetworkClient(newsocketFd);
-                
-                if (verbose == INFO)
-                    printf("server: got connection from %s port %d\n",
-                       inet_ntoa(cli_addr.sin_addr),
-                       ntohs(cli_addr.sin_port));
-
-                /* Send a welcome message */
-                sendMessage(newsocketFd, "Connection to server succeeded", "ServerStartEvent", "Server");
-
-                /* Update number of active clients */
-                activeNetworkClients++;
+                /* First check for incoming commands from network clients */
+                for (int i = 0; i < activeNetworkClients; i++)
+                {
+                    int socketIndex = ServerConstants::BASE_POLLS + i;
+                    if (pollFds[socketIndex].fd < 0)
+                        continue; /* skip connections that have been closed */
+                        
+                    if (pollFds[socketIndex].revents & (POLLERR | POLLHUP | POLLNVAL))
+                    {
+                        /* Error encountered; we should close the socket */
+                        if (pollFds[socketIndex].fd > 0)
+                        {
+                            networkClients[i]->closeFd();
+                            pollFds[socketIndex].fd = -1; /* TODO: we need a linked list rather than an array */
+                        }
+                    }
+                    else if (pollFds[ServerConstants::BASE_POLLS + activeNetworkClients].revents & POLLIN)
```

Comment:
I just noticed this post merge; shouldn't this be `socketIndex` instead of `ServerConstants::BASE_POLLS + activeNetworkClients`?

Reply:
yes. I'll correct it in another pr